sdk/core/azure-core-test/src/main/java/com/azure/core/test/utils/TestConfigurationBuilder.java

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.test.utils;
+
+import com.azure.core.implementation.util.EnvironmentConfiguration;
+import com.azure.core.util.Configuration;
+import com.azure.core.util.ConfigurationBuilder;
+import com.azure.core.util.ConfigurationSource;
+
+import java.lang.reflect.Constructor;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.junit.jupiter.api.Assertions.fail;
+
+/**
+ * Allows building test configuration including explicit and environment sources.
+ */
+public class TestConfigurationBuilder {
+    private final TestConfigurationSource source;
+    private final EnvironmentConfiguration envConfig;
+
+    /**
+     * Created {@link TestConfigurationBuilder}
+     */
+    public TestConfigurationBuilder() {
+        this.source = new TestConfigurationSource();
+        this.envConfig = new EnvironmentConfiguration();
+    }
+
+    /**
+     * Adds configuration property.
+     *
+     * @param key Property name.
+     * @param value Property value.
+     * @return the updated TestConfigurationBuilder object.
+     */
+    public TestConfigurationBuilder add(String key, String value) {
+        source.add(key, value);
+        return this;
+    }
+
+    /**
+     * Adds configuration property to environment.
+     *
+     * @param key Property name.
+     * @param value Property value.
+     * @return the updated TestConfigurationBuilder object.
+     */
+    public TestConfigurationBuilder addEnv(String key, String value) {
+        envConfig.put(key, value);
+        return this;
+    }
+
+    /**
+     * Builds shared {@link Configuration} section.
+     *
+     * @return {@link Configuration} object.
+     */
+    public Configuration build() {
+        return getBuilder().build();
+    }
+
+    /**
+     * Builds {@link Configuration} section.
+     *
+     * @param section relative section path.
+     * @return {@link Configuration} object.
+     */
+    public Configuration buildSection(String section) {
+        return getBuilder().buildSection(section);
+    }
+
+    private ConfigurationBuilder getBuilder() {
+        try {
+            Constructor<?> ctor = ConfigurationBuilder.class.getDeclaredConstructor(ConfigurationSource.class, EnvironmentConfiguration.class);
+            ctor.setAccessible(true);
+            return (ConfigurationBuilder) ctor.newInstance(source, envConfig);
```

Comment:
Thank you for pointing me to it - I didn't know `Class.newInstance` is deprecated in Java 17. Here, I use `Constructor.newInstance` which is the [official replacement](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Class.html#newInstance())

Reply:
I didn't know either :). My intention was to ask whether non-reflective solution for this exists. I'm guessing that additions in the poms like `--add-opens com.azure.core/com.azure.core.util=ALL-UNNAMED` are related to this reflective access - please correct me if this is wrong. If my suspicion is not wrong then does this mean we'd need to spread this `--add-opens` to each test package we'd like to use `TestConfigurationBuilder`? If so, then I'm a bit afraid about making it wide open in tests - might mask some access issues not visible otherwise.
I know that @alzimmermsft was doing some magic tricks with accessors in impl packages and selectively opening those to packages that required them - maybe this could work here instead of reflection.
Please let me know if this makes any sense, I'm still learning how module access works in modern Java...

Reply:
So I ended up making `ConfigurationBuilder(explicitSource, systemPropSource, envVarSource)` public. I realized even if it's not a production scenario, Spring team or any customer building configuration for them, would want to test their configuration code for SDKs and mock the environment. I.e. this API would be useful for everyone and should be public.